sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -757,6 +776,14 @@ ProgramManager::getOrCreateKernel(const ContextImplPtr &ContextImpl,
     return Cache.getOrInsertKernel(Program, KernelName);
   };
 
+  if (!SYCLConfig<SYCL_CACHE_IN_MEM>::get()) {
+    // The built kernel cannot be shared between multiple
+    // threads when caching is disabled, so we can return
+    // nullptr for the mutex.
+    auto [Kernel, ArgMask] = BuildF();
+    return make_tuple(Kernel, nullptr, ArgMask, Program);
+  }
```

Comment:
Could this logic be moved to `getOrBuild` to avoid having to do it before each call to it?

Reply:
Yes, but I think it'd probably be better suited for another PR. There are really two types of uses of `getOrBuild`, one with kernels and one with programs. The returns types differ between the two case, so it'd be a little awkward to insert this special logic for the program case in there. Alternatively, one could make a `getOrBuildProgram` and `getOrBuildKernel` program and then put that special logic in the `getOrBuildKernel` function, but there would still be a oddity with the return type: when caching, we wrap the value in a `BuildResult` and return a pointer to that (owned by the cache), but when are not caching, we bypass the `BuildResult` object, and would only want to return the value. This can still be resolved by further modifying these `getOrBuild` functions to return only the values we extract from the `BuildResult` anyways, but I believe this'll most likely create a lot of changes unrelated to the original goal of PR.

Reply:
From what I remember of the code, it is due for a bit of an overhaul anyway. 👍